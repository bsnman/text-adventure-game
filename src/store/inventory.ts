import { defineStore } from 'pinia'
import type { Item } from '../types/world'

/**
 * Pinia store managing items collected by the player.
 */
export const useInventoryStore = defineStore('inventory', {
  state: () => ({
    items: [] as Item[]
  }),
  actions: {
    /** Add an item to the inventory */
    add(item: Item) {
      this.items.push(item)
    },
    /** Remove an item by id */
    remove(id: string) {
      this.items = this.items.filter(i => i.id !== id)
    },
    /** Check if an item is currently held */
    has(id: string) {
      return this.items.some(i => i.id === id)
    }
  }
})
