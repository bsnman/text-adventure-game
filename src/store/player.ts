import { defineStore } from 'pinia'
import { calculateAttackSpeed } from '../utils/combat'

/**
 * Pinia store representing the player character.
 * Contains base stats and derived combat attributes.
 */
export const usePlayerStore = defineStore('player', {
  state: () => ({
    baseMinDamage: 2,
    baseMaxDamage: 4,
    baseAttackSpeed: 1.2,
    strength: 5,
    agility: 3,
    vitality: 4,
    intelligence: 2
  }),
  getters: {
    /** Minimum damage output including strength bonus */
    minDamage: (s) => s.baseMinDamage + s.strength,
    /** Maximum damage output including strength bonus */
    maxDamage: (s) => s.baseMaxDamage + s.strength * 1.5,
    /** Attacks per second after agility bonuses */
    attackSpeed: (s) => calculateAttackSpeed(s.baseAttackSpeed, s.agility)
  }
})
