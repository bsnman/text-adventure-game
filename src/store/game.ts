import { defineStore } from 'pinia'
import { world } from '../data/world'
import type { Coordinates, Item } from '../types/world'
import { useInventoryStore } from './inventory'

export const useGameStore = defineStore('game', {
  state: () => ({
    currentRoom: 'start',
    visited: { start: { x: 0, y: 0 } } as Record<string, Coordinates>
  }),
  getters: {
    room: (state) => world[state.currentRoom]
  },
  actions: {
    move(direction: string) {
      const next = world[this.currentRoom].exits[direction]
      if (next) {
        const offset: Record<string, [number, number]> = {
          north: [0, -1],
          south: [0, 1],
          east: [1, 0],
          west: [-1, 0]
        }
        const [dx, dy] = offset[direction] || [0, 0]
        const current = this.visited[this.currentRoom] || { x: 0, y: 0 }
        const coords = { x: current.x + dx, y: current.y + dy }
        if (!this.visited[next]) {
          this.visited[next] = coords
        }
        this.currentRoom = next
      }
    },
    /** Pick up an item from the current room and add it to the inventory */
    pickUp(id: string) {
      const room = world[this.currentRoom]
      if (!room.items) return
      const index = room.items.findIndex(i => i.id === id)
      if (index !== -1) {
        const [item] = room.items.splice(index, 1)
        useInventoryStore().add(item)
      }
    },
    /** Use an item from the inventory */
    useItem(id: string) {
      const inventory = useInventoryStore()
      if (inventory.has(id)) {
        inventory.remove(id)
      }
    }
  }
})
