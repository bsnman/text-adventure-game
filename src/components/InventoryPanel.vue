<!--
Displays the player's current inventory items.
-->
<script setup lang="ts">
import { useInventoryStore } from '../store/inventory'
const inventory = useInventoryStore()
</script>

<template>
  <v-card>
    <v-card-title>Inventory</v-card-title>
    <v-card-text>
      <ul v-if="inventory.items.length">
        <li v-for="item in inventory.items" :key="item.id">{{ item.name }}</li>
      </ul>
      <span v-else>No items</span>
    </v-card-text>
  </v-card>
</template>
