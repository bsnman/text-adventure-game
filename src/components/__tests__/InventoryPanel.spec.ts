import { describe, it, expect, vi } from 'vitest'
import { nextTick } from 'vue'
import { render } from '@testing-library/vue'
import { createTestingPinia } from '@pinia/testing'
import InventoryPanel from '../InventoryPanel.vue'
import { useInventoryStore } from '../../store/inventory'

function renderPanel() {
  const pinia = createTestingPinia({ stubActions: false, createSpy: vi.fn })
  const utils = render(InventoryPanel, { global: { plugins: [pinia] } })
  return { ...utils, pinia }
}

describe('InventoryPanel', () => {
  it('renders items from the store', async () => {
    const { getByText } = renderPanel()
    const store = useInventoryStore()
    store.add({ id: 'torch', name: 'Torch', description: '' })
    await nextTick()
    getByText('Torch')
  })

  it('matches snapshot', () => {
    const { container } = renderPanel()
    expect(container).toMatchSnapshot()
  })
})
