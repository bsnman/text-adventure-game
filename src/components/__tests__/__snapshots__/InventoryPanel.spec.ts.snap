// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InventoryPanel > matches snapshot 1`] = `
<div>
  <v-card>
    <v-card-title>
      Inventory
    </v-card-title>
    <v-card-text>
      <span>
        No items
      </span>
    </v-card-text>
  </v-card>
</div>
`;
